proxygen/lib/http/codec/compress/experimental/qcram/QCRAMDecoder.cpp

```diff
@@ -0,0 +1,296 @@
+using folly::io::Cursor;
+using std::list;
+using std::string;
+using std::unique_ptr;
+using std::vector;
+using proxygen::HPACK::DecodeError;
+
+namespace {
+const std::chrono::seconds kDecodeTimeout{5};
+}
+
+namespace proxygen {
+unique_ptr<QCRAMDecoder::headers_t> QCRAMDecoder::decode(const IOBuf* buffer) {
+  auto headers = std::make_unique<headers_t>();
+  Cursor cursor(buffer);
+  uint32_t totalBytes = buffer ? cursor.totalLength() : 0;
+  decode(cursor, totalBytes, *headers);
+  // release ownership of the set of headers
+  return headers;
+}
+
+const huffman::HuffTree& QCRAMDecoder::getHuffmanTree() const {
+  return huffman::huffTree();
+}
+
+void QCRAMDecoder::handleBaseIndex(HPACKDecodeBuffer& dbuf) {
+  if (useBaseIndex_) {
+    uint32_t baseIndex = 0;
+    err_ = dbuf.decodeInteger(8, baseIndex);
+    if (err_ != HPACK::DecodeError::NONE) {
+      LOG(ERROR) << "Decode error decoding maxSize err_=" << err_;
+      return;
+    }
+    table_.setBaseIndex(baseIndex);
+  }
+}
+
+uint32_t QCRAMDecoder::decode(Cursor& cursor,
+                              uint32_t totalBytes,
+                              headers_t& headers) {
+  uint32_t emittedSize = 0;
+  HPACKDecodeBuffer dbuf(getHuffmanTree(), cursor, totalBytes,
+                         maxUncompressed_);
+  handleBaseIndex(dbuf);
+  while (!hasError() && !dbuf.empty()) {
+    emittedSize += decodeHeader(dbuf, &headers);
+    if (emittedSize > maxUncompressed_) {
+      LOG(ERROR) << "exceeded uncompressed size limit of "
+                 << maxUncompressed_ << " bytes";
+      err_ = DecodeError::HEADERS_TOO_LARGE;
+      return dbuf.consumedBytes();
+    }
+  }
+  return dbuf.consumedBytes();
+}
+
+uint32_t QCRAMDecoder::decodeStreaming(
+    Cursor& cursor,
+    uint32_t totalBytes,
+    HeaderCodec::StreamingCallback* streamingCb) {
+
+
+  uint32_t emittedSize = 0;
+  streamingCb_ = streamingCb;
+  HPACKDecodeBuffer dbuf(getHuffmanTree(), cursor, totalBytes,
+                         maxUncompressed_);
+  handleBaseIndex(dbuf);
+  while (!hasError() && !dbuf.empty()) {
+    emittedSize += decodeHeader(dbuf, nullptr);
+
+    if (emittedSize > maxUncompressed_) {
+      LOG(ERROR) << "exceeded uncompressed size limit of "
+                 << maxUncompressed_ << " bytes";
+      err_ = HPACK::DecodeError::HEADERS_TOO_LARGE;
+      return dbuf.consumedBytes();
+    }
+  }
+
+  return dbuf.consumedBytes();
+}
+
+
+void QCRAMDecoder::handleTableSizeUpdate(HPACKDecodeBuffer& dbuf) {
+  uint32_t arg = 0;
+  err_ = dbuf.decodeInteger(5, arg);
+  if (err_ != HPACK::DecodeError::NONE) {
+    LOG(ERROR) << "Decode error decoding maxSize err_=" << err_;
+    return;
+  }
+
+  if (arg > maxTableSize_) {
+    LOG(ERROR) << "Tried to increase size of the header table";
+    err_ = HPACK::DecodeError::INVALID_TABLE_SIZE;
+    return;
+  }
+  table_.setCapacity(arg);
+}
+
+void QCRAMDecoder::updateMembership() {
+  /* Update membership */
+  uint32_t index = table_.writeBaseIndex();
```

Comment:
Just confirming, in the current QCRAM draft, this membership set is not required.  The decoder need only track the highest absolute insert index, because the decoder never inserts out of order.